### PR TITLE
New easyblock for building QuantumESPRESSO with CMake

### PR DESCRIPTION
**Pull request by Crivella**
Since Quantum Espresso 7.x,  the configure/make build has been less supported in favor of CMake/Make.

Right now bot methods of building QE coexist, but some features, like the installation of the `pw2qmcpack` are only supported in the CMake version.

Further more, the CMake implementation should be able to much more easily recognize the present external libraries and pickup environment variables set by easy build, and also properly implement parallelization in Make, allowing to build and test QE much faster (the old Config/Make would need `make` to be run multiple times to reach the end if ran with more than 1 parallel process)

### CODE REVIEW

File: easybuild/easyblocks/q/quantumespressocmake.py [hunk 1]
```diff
@@ -0,0 +1,344 @@
...
+        nprocs = self.cfg.get('test_suite_nprocs', 1)
+        self.cfg.update('configopts', '-DQE_ENABLE_TEST=ON')
+        self.cfg.update('configopts', '-DTESTCODE_NPROCS=%d' % nprocs)
+
+        # Change format of timings to seconds only (from d/h/m/s)
+        self.cfg.update('configopts', '-DQE_CLOCK_SECONDS=ON')
+
+        # Needed to avoid a DSO missing from command line linking error
+        # https://gitlab.com/QEF/q-e/-/issues/667
+        if self.cfg.get('build_shared_libs', False) :
```
[the remark this answers is not in the document]
Comment (Crivella) [reply easybuild/easyblocks/q/quantumespressocmake.py]: Fixed in f4b87a69

File: easybuild/easyblocks/q/quantumespressocmake.py [hunk 2]
```diff
@@ -0,0 +1,344 @@
...
+            raise EasyBuildError('Failed to parse test suite output')
+
+        perc = int(mch.group('perc')) / 100
+        num_fail = int(mch.group('failed'))
+        total = int(mch.group('total'))
+        passed = total - num_fail
+        failures = []  # list of tests that failed, to be logged at the end
+
+
+        # Example output for reported failures:
```
[the remark this answers is not in the document]
Comment (Crivella) [reply easybuild/easyblocks/q/quantumespressocmake.py:242]: Fixed in f4b87a69

**Comment by Crivella** [comment]
One thing to note is that compilation of some QE components and plugins has been moved to an implementantion with `submodule` pulling data from a specific commit
Offline building is still possible but will require more careful implementation, at least if we still want to check the hash, while running the script as suggested [on the QE wiki](https://gitlab.com/QEF/q-e/-/wikis/Support/Build-without-internet).

While it is possible to emulate what the script is doing, i would prefer to keep it as a source of truth, in case paths/implementation are changed in future versions.

File: easybuild/easyblocks/q/quantumespresso.py [hunk 3]
```diff
@@ -367,6 +367,12 @@ def _adjust_compiler_flags(self, comp_fam):
     def configure_step(self):
         """Custom configuration procedure for Quantum ESPRESSO."""
 
+        if LooseVersion(self.version) < LooseVersion("7.3.1"):
```
Comment (ocaisa) [inline easybuild/easyblocks/q/quantumespresso.py:370]:
```suggestion
        if LooseVersion(self.version) >= LooseVersion("7.3.1"):
```

**Review by ocaisa — CHANGES_REQUESTED** [review]

**Review by ocaisa — APPROVED** [review]
LGTM

**Comment by ocaisa** [comment]
@boegelbot please test @ jsc-zen3
EB_ARGS="QuantumESPRESSO-7.3-foss-2023a.eb"

**Comment by Crivella** [comment]
This flakiness with the thresholds on the test-suite...

**Comment by ocaisa** [comment]
There is a PR open that addresses the single failing test issue that we have seen for QE in EESSI for the original easyblock. The reason for running that build here was just a sanity check, and that was ok.

In https://github.com/easybuilders/easybuild-easyconfigs/pull/20138 we have the checks for the new easyblock.

**Comment by ocaisa** [comment]
@boegelbot please test @ jsc-zen3

File: easybuild/easyblocks/q/quantumespressocmake.py [hunk 4]
```diff
@@ -0,0 +1,405 @@
...
+from easybuild.framework.easyconfig import CUSTOM
+from easybuild.tools import LooseVersion
+from easybuild.tools.build_log import EasyBuildError
+from easybuild.tools.modules import get_software_root
+from easybuild.tools.run import run_cmd
+
+from easybuild.easyblocks.generic.cmakemake import CMakeMake
+
+
+class EB_QuantumESPRESSOcmake(CMakeMake):
```
[the remark this answers is not in the document]
Comment (ocaisa) [reply easybuild/easyblocks/q/quantumespressocmake.py:44]: I had a different plan, I was going to do a follow-up PR when the open QE easyconfigs are merged, where we use an explicit easyblock for the configuremake-based easyconfigs and then switch this to the standard

[the remark this answers is not in the document]
Comment (ocaisa) [reply easybuild/easyblocks/q/quantumespressocmake.py:44]: It would be a lot of (pointless?) effort to merge them

**Pull request merged**